File: src/py/omni_token_economy/redundancy.py
"""Redundancy detection via asymmetric word overlap."""
from __future__ import annotations

import re

_WORD_RE = re.compile(r"[^\W_]+", re.UNICODE)


def _words(s: str) -> set[str]:
    return set(_WORD_RE.findall(s.lower()))


def detect_redundancy(a: str, b: str) -> float:
    """Return |words(a) ∩ words(b)| / |words(a)|. 0.0 when either empty.

    Asymmetric on purpose — measures how much of `a` is covered by `b`.
    """
    if not a or not b:
        return 0.0
    a_low = a.lower().strip()
    b_low = b.lower().strip()
    if a_low == b_low:
        return 1.0
    if a_low in b_low:
        return 1.0
    wa = _words(a_low)
    if not wa:
        return 0.0
    wb = _words(b_low)
    inter = len(wa & wb)
    return inter / len(wa)


def is_redundant(short: str, long: str, threshold: float = 0.6) -> bool:
    """True if `short` is covered by `long` above threshold."""
    return detect_redundancy(short, long) >= threshold

File: src/ts/redundancy.ts
const WORD_RE = /[\p{L}\p{N}]+/gu;

function words(s: string): Set<string> {
  return new Set((s.toLowerCase().match(WORD_RE) ?? []));
}

/**
 * Word overlap ratio: |A ∩ B| / |A|.
 * Asymmetric on purpose — measures how much of `a` is covered by `b`.
 * Returns 0 when either is empty.
 */
export function detectRedundancy(a: string, b: string): number {
  if (!a || !b) return 0;
  const aLow = a.toLowerCase().trim();
  const bLow = b.toLowerCase().trim();
  if (aLow === bLow) return 1;
  if (bLow.includes(aLow)) return 1;
  const wa = words(aLow);
  const wb = words(bLow);
  if (wa.size === 0) return 0;
  let inter = 0;
  for (const w of wa) if (wb.has(w)) inter++;
  return inter / wa.size;
}

/**
 * True if `short` can be considered redundant given `long`.
 * Uses detectRedundancy >= threshold.
 */
export function isRedundant(short: string, long: string, threshold = 0.6): boolean {
  return detectRedundancy(short, long) >= threshold;
}
